pkg/discovery/processor/cluster_processor.go

```diff
@@ -198,3 +271,52 @@ func (p *ClusterProcessor) DiscoverCluster() (*repository.ClusterSummary, error)
 
 	return repository.CreateClusterSummary(kubeCluster), nil
 }
+
+// calculateItemsPerListQuery dynamically calculates the number of items per query to avoid OOM.
+// This value must be calculated dynamically because:
+//   - the number of pods changes over time
+//   - the kubeturbo memory limit can change over time (when in-place pod resize is enabled)
+//
+// The following formula is used:
+//
+// (kubeturbo_mem_limit_gb * default_automemlimit
+//   - generic_metric_size_gb_per_pod * number_of_pods_in_thousands_in_cluster
+//   - (cadvisor_metric_size_gb_per_pod * max_pods_per_node + extra_per_node_usage_gb)
+//   - extra_cluster_wide_usage_gb) * items_per_gb
+//
+// where the following values are set based on kubeturbo memory profiling:
+//   - default_automemlimit = 0.9 (percentage of memory limit that can be controlled by Go Runtime)
+//   - generic_metric_size_gb_per_pod = 0.24
+//   - cadvisor_metric_size_gb_per_pod = 0.004
+//   - max_pods_per_node = 250
+//   - extra_per_node_usage_gb = 0.2
+//   - extra_cluster_wide_usage_gb = 0.2
+//   - items_per_gb = 5000
+//
+// which equals to:
+//
+//	(kubeturbo_mem_limit_gb * 0.9 - number_of_pods_in_thousands_in_cluster * 0.24 - 1.4) *  5000
+func (p *ClusterProcessor) calculateItemsPerListQuery(podCount int) (int, float64, error) {
+	limit, err := memlimit.FromCgroup()
+	if err != nil {
+		// This is very unlikely because in absense of any limit set in container resources
+		// we will get the cgroup limit as the nodes available/usable memory limit
+		return 0, 0, fmt.Errorf("error retrieving memory limit (%v): %v", limit, err)
+	}
+	if limit == 0 {
+		// This is very unlikely because in absense of any limit set in container resources
+		// we will get the cgroup limit as the nodes available/usable memory limit
+		return 0, 0, fmt.Errorf("limit found set to zero (0")
+	}
+	podsInThousands := float64(podCount) / 1000
+	limitInGB := nodeUtil.Base2BytesToGigabytes(float64(limit))
+	availMem := limitInGB*DefaultAutoMemLimitPct -
+		DefaultGenericMetricSizePerThousandPodsInGB*podsInThousands -
+		(DefaultCadvisorMetricSizePerPodInGB*DefaultMaxPodsPerNode + DefaultExtraPerNodeUsageInGB) -
+		DefaultExtraClusterWideUsageInGB
+	if availMem < 1.0 {
+		return 0, 0, fmt.Errorf("unable to calculate reasonable items per list query based on "+
```

Comment:
While I agree that it will be too rough an estimate when `availMem` becomes too small, it's still probably better than setting it to the default (which is 27800 and is much bigger than 5000).  I'd suggest we consider the followings:
- I think we can set a minimum, e.g., 2000 items per query (instead of the default which is 27800).
- If `availMem` is larger than 0.4 for example, we still use `DefaultItemsPerGBMemory * availMem`.
- If smaller, then we use the minimum instead of the default.  In the log message, we should say this represents not enough memory allocated for kubeturbo and one should increase it.  Furthermore, we can estimate how many round trips it would take and warn the user that it could take a long time to discover and it could also run into OOM.

Reply:
How about further simplify this, and get rid of the `DefaultItemsPerListQuery` of 27800, and just make the default to 5000?

* If `availMem` is smaller than 1, use  `DefaultItemsPerGBMemory` (5000), and show logs to suggest increase the memory limit.
* Otherwise, use `DefaultItemsPerGBMemory * availMem`

Reply:
This has been agreed.